ts/src/profiler.ts

```diff
@@ -201,29 +254,37 @@ export class Profiler extends common.ServiceObject {
     const delayMillis = await this.collectProfile();
 
     // Schedule the next profile.
-    setTimeout(this.runLoop.bind(this), delayMillis).unref();
+    if (typeof delayMillis === 'number') {
+      setTimeout(this.runLoop.bind(this), delayMillis).unref();
+      return;
+    }
+
+    setTimeout(this.runLoop.bind(this), delayMillis.getBackoff()).unref();
   }
 
   /**
    * Waits for profiler server to tell it to collect a profile, then collects
    * a profile and uploads it.
    *
    * @return - time, in ms, to wait before asking profiler server again about
-   * collecting another profile.
-   *
-   * TODO: implement backoff and retry. When error encountered in
-   * createProfile() should be retried when response indicates this request
-   * should be retried or with exponential backoff (up to one hour) if the
-   * response does not indicate when to retry this request.
+   * collecting another profile, or retryer specifying exponential backoff.
    */
-  async collectProfile(): Promise<number> {
+  async collectProfile(retryer?: Retryer): Promise<number|Retryer> {
     let prof: RequestProfile;
     try {
       prof = await this.createProfile();
     } catch (err) {
       this.logger.error(
           `Error requesting profile type to be collected: ${err}.`);
-      return this.config.backoffMillis;
+      if (err.backoffMillis !== undefined) {
```

Comment:
I think the code would be cleaner if for the server-instructed backoff a separate type of error would thrown and you would check for the type (and for that type the backoff duration field would always be set, as an invariant).

Reply:
Done.